---
title: "DUX4 recruits p300/CBP through its C-terminus and induces global H3K27 acetylation changes"
author: "micah gearhart"
date: "10/11/2015"
output:
  md_document:
    variant: markdown_github
---

```{r libraries,eval=T,results='hide',message=FALSE,warning=FALSE}
library("DESeq2")
library("ggplot2")
library("gridExtra")
library("pheatmap")
library("BiocParallel")
library("GenomicAlignments")
library("Rsamtools")
library("rtracklayer")
library("RColorBrewer")
library("dplyr")
library("tidyr")
library(BSgenome.Hsapiens.UCSC.hg19)
hg19<-BSgenome.Hsapiens.UCSC.hg19
library("ChIPpeakAnno")
source("R/hello.R")

ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

#Count the Number or Reads in Each ChIP Dataset
```{r Filesets,eval=F}
h3k27<-new("fileset", filename=c( "../chip/h3k27_nodox.R1_trimmed.fastq.hg19.bam",
                                  "../chip/h3k27_plusdox.R1_trimmed.fastq.hg19.bam"),
           labels=c("H3K27ac -Dox","H3K27ac +Dox"))
(h3k27<-countFileset(h3k27))

DUX4i<-new("fileset",filename=c("../chip/input_plusdox.R1_trimmed.fastq.hg19.bam",
                               "../chip/DUX4.R1_trimmed.fastq.hg19.bam"),
           labels=c("Input +Dox","Flag +Dox"))
(DUX4i<-countFileset(DUX4i))

#pan Histone H3
H3<-new("fileset", filename=c( "../chip/h3_nodox.R1_trimmed.fastq.hg19.bam",
                                  "../chip//h3_plusdox.R1_trimmed.fastq.hg19.bam"),
           labels=c("H3 -Dox","H3 +Dox"))
(H3<-countFileset(H3))

#H3K4me3
h3k4<-new("fileset",filename=c("../chip/h3k4me_s91_nodox.R1_trimmed.fastq.hg19.bam",
           "../chip/h3k4me_s92_plusdox.R1_trimmed.fastq.hg19.bam"),
           labels=c("H3K4me3 -Dox","H3K4me3 +Dox"))
(h3k4<-countFileset(h3k4))

#ENCODE DATA
dnaseFS<-new("fileset",filename=c("../sra/ENCFF001BUR.bam"), labels=c("ENCODE DNAse"))
(dnaseFS<-countFileset(dnaseFS))

save(h3k27,H3,DUX4i,h3k4,dnaseFS,file="filesets.rdata")
```

# Import Peaklist To Use
``` {r Peaklists,eval=T}
#hg19blacklist
#downloader::download("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeDacMapabilityConsensusExcludable.bed.gz",destfile="hg19_blacklist.bed.gz")
hg19bl<-rtracklayer::import("hg19_blacklist.bed")

dux4_tap<-rtracklayer::import("../sra/dux4_fl_pe5_peaks.bed")
dux4_tap<-keepSeqlevels(dux4_tap,seqlevels(hg19)[1:24])
length(dux4_tap<-dux4_tap[!dux4_tap %over% hg19bl])

dux4_vs_input<-rtracklayer::import("../chip/DUX4dox_pe5_peaks.bed")
dux4_vs_input_summits<-rtracklayer::import("../chip/DUX4dox_pe5_summits.bed")

#dux4_vs_input<-keepSeqlevels(dux4_vs_input,seqlevels(hg19)[1:24])
#length(dux4_vs_input<-dux4_vs_input[!dux4_vs_input %over% hg19bl])

#Import H327Ac Called Peaks
length(k27pd<-import("../chip/h3k27ac_plusdox_pe5_peaks.bed"))
length(k27nd<-import("../chip/h3k27ac_nodox_pe5_peaks.bed"))

#Import DNAse data from ENCODE
#downloader::download("https://www.encodeproject.org/files/ENCFF001BVR/@@download/ENCFF001BVR.bigBed",destfile="ENCFF001BVR.bigBed")
dnase<-read.table("ENCFF001BVR.bed",stringsAsFactors=FALSE)
dnase<-GRanges(seqnames=dnase$V1,IRanges(start=dnase$V2,end=dnase$V3),score=dnase$V7,score2=dnase$V8)

#Download Published Data from http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE33838
temp<-read.csv("GSM837613_Dux4.full.peak.csv",stringsAsFactors=F)
tsGR<-GRanges(seqnames=temp$chr,IRanges(start=temp$start,end=temp$end),max.cov=temp$max.cov,pval=temp$pval)

hg18ToHg19<-import.chain("hg18ToHg19.over.chain")
length(tsGR_hg19<-unlist(liftOver(tsGR,hg18ToHg19)))
#export(tsGR_hg19,"tsGR_hg19.bed")
```

#Count Reads under DUX4 Peaks in ChIP datasets
```{r Count_Reads_DUX4,eval=F}
dux4dox_1k<-dux4_vs_input_summits+500
dux4dox_4k<-dux4_vs_input_summits+2000

(fls <- list.files("../chip", pattern=glob2rx("*hg19.bam$"),full=TRUE))
bamlst <- BamFileList(fls,yieldSize = 1e5)
detectCores()
BiocParallel::register(MulticoreParam(workers=detectCores()))
system.time(dux4dox_1k_counts <- summarizeOverlaps(dux4dox_1k,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE))
system.time(dux4dox_4k_counts <- summarizeOverlaps(dux4dox_4k,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE))

apply(assays(dux4dox_1k_counts)$counts,2,sum)
apply(assays(dux4dox_4k_counts)$counts,2,sum)
save(dux4dox_1k_counts,dux4dox_4k_counts,file="DUX4_counts.rdata")
```

# Evaluate Counts under DUX4 Peaks
```{r violin_dux4_sites,eval=T}
load("filesets.rdata")
load("DUX4_counts.rdata")

head(assays(dux4dox_1k_counts)$counts)
gr<-rowRanges(dux4dox_1k_counts)
head(gr)
#idx<-!keepSeqlevels(gr,seqlevels(hg19)[1:24]) %over% hg19bl
#allcounts<-assays(dux4dox_1k_counts)$counts[idx,]
allcounts<-assays(dux4dox_1k_counts)$counts[,]
#(n<-colSums(allcounts))
#length(gr<-gr[idx])
gr$k27nd<-log2(allcounts[,2]*1e6/h3k27@count[1] + 1)
gr$k27pd<-log2(allcounts[,3]*1e6/h3k27@count[2] + 1)
gr$input<-log2(allcounts[,7]*1e6/DUX4i@count[1] + 1)
gr$dux4<-log2(allcounts[,1]*1e6/DUX4i@count[2] + 1)

#Filter ranges on short Contigs
length(gr<-keepSeqlevels(gr,seqlevels(hg19)[1:24]))

#Filter ranges that overlap BlackListed Peaks
length(gr<-gr[!gr %over% hg19bl])

#Filter ranges with 'outlier' input counts
(cutoff <-quantile(as.numeric(gr$input), 3/4)+1.5*IQR(as.numeric(gr$input)))
length(gr<-gr[gr$input < cutoff])

(gg_violin<- as.data.frame(mcols(gr)[,3:4]) %>%
  gather(condition,log2cpm) %>%
#  dplyr::filter(log2cpm > 0) %>%
  ggplot(aes(x=condition,y=log2cpm)) + ggtitle("K27ac counts at DUX4 Sites") +
   xlab("Condition") + ylab("Log2(cpm)") +
   geom_violin(aes(fill=condition)) +
  scale_fill_manual(values=cbPalette[c(7,6)]) +
   theme_bw() + theme(panel.grid.major=element_blank(),
                     panel.grid.minor=element_blank())
)

#ggsave(file="011916_violin.svg",device = svglite::svglite,plot=gg_violin,width=8.5,height=6)
save(gg_violin,file=paste0(ts,"_figure5b.rdata"))

#Order by DUX4 peak score
gr<-gr[with(gr,order(-score))]

#Add a logical to indicated DNAse Accessible Overlapping 1kb window   
gr$dnase_overlap <- overlapsAny(gr,dnase,maxgap=0L)

#Replace dux4dox_1k with "cleaned up" version
length(dux4dox_1k <- gr)
length(dux4dox_1k_dnase <- gr[gr$dnase_overlap])
length(dux4dox_1k_nodnase <- gr[!gr$dnase_overlap])

#Also Filter full Length DUX4 Peaks
length(dux4_vs_input<-dux4_vs_input[dux4_vs_input$name %in% dux4dox_1k$name])
seqlevels(dux4_vs_input) <- seqlevels(dux4dox_1k)
length(dux4_vs_input) == length(dux4dox_1k)

length(dux4_vs_input_summits<-dux4_vs_input_summits[dux4_vs_input_summits$name %in% dux4dox_1k$name])
seqlevels(dux4_vs_input_summits) <- seqlevels(dux4dox_1k)
length(dux4_vs_input_summits) == length(dux4dox_1k)

```

# Measure Overlap with Geng et al 2012 Data
```{r Geng_Overlap,eval=T}
#Overlap with Published Peak List
mean(dux4_vs_input  %over% tsGR_hg19)

#Overlap with  MACS calls of Geng et al  DUX4 ChIP data
mean(dux4_vs_input %over% dux4_tap)
```
 
#Figure 5c
```{r Dux4_h3k27ac_DUX4sites,eval=F}
benchplot(dux4_total<-twister(dux4dox_1k,dataset=DUX4i,pad = 3500,ord=0,window=1,ya=c(12,16)))
benchplot(h3k27ac_total<-twister(dux4dox_1k,dataset=h3k27,pad = 3500,ord=0,window=1,ya=c(12,16)))
grid.arrange(dux4_total,h3k27ac_total,ncol=1)
save(dux4_total,h3k27ac_total,file=paste0(ts,"_figure5c.rdata"))
```


# Estimate Overlaps with LTR Elements
```{r maLR_new,eval=T}
load("rmsk_full_hg19.rdata")
# tbl.rmsk<-tbl.rmsk[tbl.rmsk$repFamily=="ERVL-MaLR",]
maLR<-GRanges(seqnames=tbl.rmsk$genoName,ranges=IRanges(start=tbl.rmsk$genoStart,end=tbl.rmsk$genoEnd),
              strand=tbl.rmsk$strand,name=tbl.rmsk$repName,family=tbl.rmsk$repFamily,class=tbl.rmsk$repClass)
maLR<-keepSeqlevels(maLR,seqnames(hg19)[1:24])
 
m<-maLR[maLR$class == "LTR"]

temp<-table(m$family)
m$family<-as.factor(gsub("\\?","",as.character(m$family)))
m<-m[m$family %in% names(temp)[temp>0]]
m$name<-factor(m$name) #drop levels in an mcol
m$family<-factor(m$family) #drop levels in an mcol
table(m$family)
mL<-split(m,m$family)

fx2 <- function(repeats,peaks) { sum(overlapsAny(peaks,repeats,ignore.strand=TRUE)) }

ltr_overlaps<- function(d4) {
x<-cbind(dnase=unlist(lapply(mL,fx2,d4[d4 %over% dnase])),
      nodnase=unlist(lapply(mL,fx2,d4[!d4 %over% dnase])))
x1<-sum(!d4[d4 %over% dnase] %over% m)
x2<-sum(!d4[!d4 %over% dnase] %over% m)
x<-rbind(x,none=c(x1,x2))
return(x)
}

ltr_overlaps(dux4_vs_input_summits)
```

# Make Pie Graph Showing MaLR overlaps
```{r maLR_pie,eval=T}
dux4_vs_input_summits$dnase<-ifelse(overlapsAny(dux4_vs_input_summits,dnase),"DNAse Accessible","DNAse Inaccessible")
dux4_vs_input_summits$malr<-ifelse(overlapsAny(dux4_vs_input_summits,mL[["ERVL-MaLR"]]),"+ MaLR","- MaLR")
dux4_vs_input_summits$category<-paste0(dux4_vs_input_summits$dnase," ",dux4_vs_input_summits$malr)

pie<-as.data.frame(mcols(dux4_vs_input_summits)) %>% 
mutate(category=factor(category,
                       levels=c("DNAse Accessible - MaLR",
                                "DNAse Accessible + MaLR",
                                "DNAse Inaccessible - MaLR",
                                "DNAse Inaccessible + MaLR" ))) %>% 
ggplot(aes(x=factor(1),fill=category)) +geom_bar(width=1)+coord_polar(theta="y")+
  scale_fill_manual(values=cbPalette[c(3,6,2,7)],guide = guide_legend(reverse=FALSE))+
  xlab("") + ylab("") +
  theme_bw() + theme(panel.grid.major=element_blank(),
                     panel.grid.minor=element_blank())

table(dux4_vs_input_summits$category)

#ggsave(file="012816_pie.svg",device = svglite::svglite,plot=pie,width=8.5,height=6)
save(pie,file=paste0(ts,"_figure5e_pie.rdata"))

```


```{r maLR_ERV,eval=F}
#mySession = browserSession("UCSC")
#genome(mySession) <- "hg19"
#tbl.rmsk <- getTable(ucscTableQuery(mySession, track="rmsk",table="rmsk"))
#save(tbl.rmsk,file="rmsk_full_hg19.rdata")
load("rmsk_full_hg19.rdata")
 tbl.rmsk<-tbl.rmsk[tbl.rmsk$repFamily=="ERVL-MaLR",]
 maLR<-GRanges(seqnames=tbl.rmsk$genoName,ranges=IRanges(start=tbl.rmsk$genoStart,end=tbl.rmsk$genoEnd),strand=tbl.rmsk$strand,name=tbl.rmsk$repName)
 maLR<-keepSeqlevels(maLR,seqnames(hg19)[1:24])
 maLR <- sortSeqlevels(maLR)
 maLR<-sort(maLR,ignore.strand=TRUE)
 #export(maLR,"malR_hg19.bed")

#DNAse
length(dux4dox_1k_dnase_maLR <- dux4dox_1k_dnase[dux4dox_1k_dnase %over% maLR])
length(dux4dox_1k_dnase_maLR)/length(dux4dox_1k_dnase)

length(dux4dox_1k_dnase_nomaLR <- dux4dox_1k_dnase[!dux4dox_1k_dnase %over% maLR])
length(dux4dox_1k_dnase_nomaLR)/length(dux4dox_1k_dnase)

#NoDNAse
length(dux4dox_1k_nodnase_maLR <- dux4dox_1k_nodnase[dux4dox_1k_nodnase %over% maLR])
length(dux4dox_1k_nodnase_maLR)/length(dux4dox_1k_nodnase)

length(dux4dox_1k_nodnase_nomaLR <- dux4dox_1k_nodnase[!dux4dox_1k_nodnase %over% maLR])
length(dux4dox_1k_nodnase_nomaLR)/length(dux4dox_1k_nodnase)
```

# MaLR H3K4me3
```{r h3k4me3_distribution_malR,eval=F,fig.height=10,fig.width=6}
h3k4me3_dnase <-twister(dux4dox_1k_dnase,dataset=h3k4,pad = 3500,ord=0,window=1,color="blue",ya=c(10,16))
h3k4me3_nodnase_maLR<-twister(dux4dox_1k_nodnase_maLR,dataset=h3k4,pad = 3500,ord=0,window=1,color="blue",ya=c(10,16))
h3k4me3_nodnase_nomaLR<-twister(dux4dox_1k_nodnase_nomaLR,dataset=h3k4,pad = 3500,ord=0,window=1,color="blue",ya=c(10,16))
save(h3k4me3_dnase,h3k4me3_nodnase_maLR,h3k4me3_nodnase_nomaLR,file=paste0(ts,"_figure5d_h3k4_malr.rdata"))
grid.arrange(h3k4me3_dnase,h3k4me3_nodnase_maLR,h3k4me3_nodnase_nomaLR,ncol=1)
```

# MaLR H3K7Ac
```{r h3k27ac_distribution_malR, eval=F,fig.height=10,fig.width=6}
h3k27_dnase <-twister(dux4dox_1k_dnase,dataset=h3k27,pad = 3500,ord=0,window=1,color="blue",ya=c(8,16))
h3k27_nodnase_maLR<-twister(dux4dox_1k_nodnase_maLR,dataset=h3k27,pad = 3500,ord=0,window=1,color="blue",ya=c(8,16))
h3k27_nodnase_nomaLR<-twister(dux4dox_1k_nodnase_nomaLR,dataset=h3k27,pad = 3500,ord=0,window=1,color="blue",ya=c(8,16))
save(h3k27_dnase,h3k27_nodnase_maLR,h3k27_nodnase_nomaLR,file=paste0(ts,"_figure5d_h3k27_malr.rdata"))
grid.arrange(h3k27_dnase,h3k27_nodnase_maLR,h3k27_nodnase_nomaLR,ncol=1)
```

### Figure 5D

# DUX4 
```{r dux4_distribution,eval=F,fig.height=10,fig.width=6}
dux4_dnase <-twister(dux4dox_1k_dnase,dataset=DUX4i,pad = 3500,ord=0,window=1,color="blue",ya=c(13,16))
dux4_nodnase<-twister(dux4dox_1k_nodnase,dataset=DUX4i,pad = 3500,ord=0,window=1,color="blue",ya=c(13,16))
save(dux4_dnase,dux4_nodnase,file=paste0(ts,"_figure5d_dux4.rdata"))
grid.arrange(dux4_dnase,dux4_nodnase,ncol=1)
```

# H3
```{r h3_distribution,eval=F,fig.height=10,fig.width=6}
h3_dnase <-twister(dux4dox_1k_dnase,dataset=H3,pad = 3500,ord=0,window=1,color="blue",ya=c(13,15))
h3_nodnase<-twister(dux4dox_1k_nodnase,dataset=H3,pad = 3500,ord=0,window=1,color="blue",ya=c(13,15))
save(h3_dnase,h3_nodnase,file=paste0(ts,"_figure5d_h3.rdata"))
grid.arrange(h3_dnase,h3_nodnase,ncol=1)
```

# H3K27Ac
```{r h3k27ac_distribution,eval=F,fig.height=10,fig.width=6}
h3k27ac_dnase <-twister(dux4dox_1k_dnase,dataset=h3k27,pad = 3500,ord=0,window=1,color="blue",ya=c(12,16))
h3k27ac_nodnase<-twister(dux4dox_1k_nodnase,dataset=h3k27,pad = 3500,ord=0,window=1,color="blue",ya=c(12,16))
save(h3k27ac_dnase,h3k27ac_nodnase,file=paste0(ts,"_figure5d_h3k27.rdata"))
grid.arrange(h3k27ac_dnase,h3k27ac_nodnase,ncol=1)
```

# H3K4me3
```{r h3k4me3_distribution,eval=F,fig.height=10,fig.width=6}
h3k4me3_dnase <-twister(dux4dox_1k_dnase,dataset=h3k4,pad = 3500,ord=0,window=1,color="blue",ya=c(11,16))
h3k4me3_nodnase<-twister(dux4dox_1k_nodnase,dataset=h3k4,pad = 3500,ord=0,window=1,color="blue",ya=c(11,16))
save(h3k4me3_dnase,h3k4me3_nodnase,file=paste0(ts,"_figure5d_h3k4.rdata"))
grid.arrange(h3k4me3_dnase,h3k4me3_nodnase,ncol=1)
```


### Analyze K27ac at non-DUX4 Sites
```{r h27ac_2k,eval=F}
length(k27pdr<-reduce(k27pd,min.gapwidth=1000))
length(k27ndr<-reduce(k27nd,min.gapwidth=1000))
length(k27r<-reduce(c(k27pdr,k27ndr),min.gapwidth=1))
k27r<-keepSeqlevels(k27r,seqlevels(hg19)[1:24])
mean(idx<-!k27r %over% hg19bl)
k27r<-k27r[idx]
summary(width(k27r))
summary(width(center(k27r)+1000))
k27rc_2k<-center(k27r)+1000

length(dux4_vs_input)
length(k27rc_2k_noDux<-k27rc_2k[!(k27rc_2k+10000) %over% dux4_vs_input])
(fls <- list.files("../chip", pattern=glob2rx("h3k27*hg19.bam$"),full=TRUE))
bamlst <- BamFileList(fls,yieldSize = 1e5)
detectCores()
BiocParallel::register(MulticoreParam(workers=detectCores()))
system.time(h3k27_counts <- summarizeOverlaps(k27rc_2k_noDux,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE))

n<-apply(assays(h3k27_counts)$counts,2,sum)
x<-1e6*assays(h3k27_counts)$counts
x[,1]<-x[,1]/h3k27@count[1]
x[,2]<-x[,2]/h3k27@count[2]

x<-log2(x+1)

(gg_violin2<-as.data.frame(x) %>% 
  gather(condition,log2cpm) %>% 
  #dplyr::filter(log2cpm > 0) %>%
  ggplot(aes(x=condition,y=log2cpm)) + ggtitle("K27ac counts at Non-DUX4 Sites") +
   xlab("Condition") + ylab("Log2(cpm)") + 
   geom_violin(aes(fill=condition)) + 
  scale_fill_manual(values=cbPalette[c(7,6)]) +
   theme_bw() + theme(panel.grid.major=element_blank(),
                     panel.grid.minor=element_blank()))

save(gg_violin2,file=paste0(ts,"_figure6a_violin2.rdata"))
```

# K27Ac Tornado Plots

```{r k27ac_tornado,eval=F,fig.height=14,fig.width=6}
temp<-k27rc_2k_noDux
temp$ratio<-x[,2]-x[,1]
temp<-temp[with(temp,order(-ratio))]

k27r_k27_dnase<-tornado(temp,dataset=h3k27,pad = 4500,ord=0,window=5,color="darkgreen")
ggsave(file=paste0(ts,"_figure6b_k27ac.png"),plot=k27r_k27_dnase,width=5,height=8)

k27r_DUX4i_dnase<-tornado(temp,dataset=DUX4i,pad = 4500,ord=0,window=5,color="blue")
ggsave(file=paste0(ts,"_figure6a_violin2.png"),plot=k27r_DUX4i_dnase,width=5,height=8)

grid.arrange(k27r_k27_dnase,k27r_DUX4i_dnase, ncol=1)
```

### RNA-Seq Analysis
```{r count_hits,eval=F}
#Use most recent hg19 build
ensembl_75 = useMart(host = "feb2014.archive.ensembl.org", biomart = "ENSEMBL_MART_ENSEMBL", 
    dataset = "hsapiens_gene_ensembl")        
        
txdb<-makeTxDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",host = "feb2014.archive.ensembl.org",dataset="hsapiens_gene_ensembl")
ens75<-exonsBy(txdb,by="gene")
save(ens75,file="ens75.rdata")

(fls <- list.files(".", pattern="GRCh37.q4.bam$",full=TRUE))
register(MulticoreParam(workers=12))
load("ens75.rdata")
bamlst <- BamFileList(fls)
hits_q4 <- summarizeOverlaps(ens75,bamlst,mode="Union",singleEnd=FALSE,ignore.strand=TRUE)
apply(assays(hits_q4)$counts,2,sum)

(fls <- list.files(".", pattern="GRCh37.bam$",full=TRUE))
bamlst <- BamFileList(fls)
hits <- summarizeOverlaps(ens75,bamlst,mode="Union",singleEnd=FALSE,ignore.strand=TRUE)
apply(assays(hits)$counts,2,sum)
save(hits_q4,hits,file="GRCh37_hits.rdata")
```

```{r DESEQ2,eval=T}
load("../rna/GRCh37_hits.rdata")
load("hgnc.rdata")

cds<-DESeqDataSet(hits_q4,design=~1)
colnames(cds)<-sapply(strsplit(colnames(cds),"_"),function(x) x[1])
cds$dox<-as.factor(rep(c("nodox","plusdox"),each=3))
design(cds)<-(~dox)

plotPCA( DESeqTransform( cds ) ,intgroup = c("dox"))+
  ggtitle("Dux4 Inducible RNA-Seq") + theme_bw()
cds<-DESeq(cds)
plotMA(cds,ylim=c(-8,8))

#results
res<-results(cds)
res<-as.data.frame(res)

#get annotations
#ensembl_83 = useMart(biomart="ENSEMBL_MART_ENSEMBL", host="www.ensembl.org", path="/biomart/martservice",dataset="hsapiens_gene_ensembl")
#hgnc<-getBM(filters="ensembl_gene_id",values=rownames(res),
#      attributes=c("ensembl_gene_id","hgnc_symbol"),mart=ensembl_83)
#save(hgnc,file="hgnc.rdata")
idx<-match(rownames(res),hgnc$ensembl_gene_id)
res$hgnc<-hgnc[idx,"hgnc_symbol"]
res<-res[!is.na(res$padj),]
res<-res[with(res,order(padj,-log2FoldChange)),]
res[grep("ZSCAN4",res$hgnc),]

#plot(res$log2FoldChange,-1*log10(res$padj),cex=0.5,pch=16)


#11-17-15
f<-as.data.frame(fpkm(cds))
idx<-match(rownames(f),hgnc$ensembl_gene_id)
f$hgnc<-hgnc[idx,2]
head(f)
f[grep("ZSCAN4",f$hgnc),]
f[grep("MYOD1",f$hgnc),]
write.csv(f,file="Supplementary_Table_1.csv",quote=F)
```

```{r volcano_plot2,eval=T,fig.height=10,fig.width=6}
#tss<-getAnnotation(ensembl_75,featureType="TSS",output="GRanges")
load("tss.rdata")
length(dux4_vs_input)
dux4_vs_input_anno <- annotatePeakInBatch(dux4_vs_input, AnnotationData=tss, output="nearest", maxgap=100L)
summary(dux4_vs_input_anno$shortestDistance)
#res$peak10kb<-rownames(res) %in% unique(dux4_vs_flag_anno$feature)
res$peak10kb<-"> 50kb"
overs<-unique(dux4_vs_input_anno[dux4_vs_input_anno$insideFeature=="overlapStart",]$feature)
fivekb<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 5000,]$feature)
fivekb<-fivekb[!fivekb %in% overs]
tenkb<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 10000,]$feature)
tenkb<-tenkb[!(tenkb %in% fivekb) & !(tenkb %in% overs)]
fiftykb<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 50000,]$feature)
fiftykb<-tenkb[!(fiftykb %in% tenkb) & !(fiftykb %in% fivekb) & !(fiftykb %in% overs)]
res[rownames(res) %in% overs,"peak10kb"]<-"Overlap's Start"
res[rownames(res) %in% fivekb,"peak10kb"]<-"< 5Kb"
res[rownames(res) %in% tenkb,"peak10kb"]<-"< 10Kb"
res[rownames(res) %in% fiftykb,"peak10kb"]<-"< 50Kb"
table(res$peak10kb)

gg1<-res %>% 
  mutate(mlogp=-1*log10(padj)) %>% 
  mutate(peak10kb=factor(peak10kb,levels=c("Overlap's Start","< 5Kb","< 10Kb","< 50Kb","> 50kb"))) %>% 
  ggplot(aes(x=log2FoldChange,y=mlogp,color=peak10kb)) +
  geom_point(size=3) + scale_color_manual(values=cbPalette[c(2,6,7,8,1)],
                                    name="Distance to Dux4 Peak") +
  xlab("DUX4 Induced Log2 Fold Change")+
  ggtitle("All Genes")+
  ylab("-log10(Adjusted P Value)") +
  xlim(c(-3,7)) + ylim(c(0,300)) + 
  theme_bw() + theme(panel.grid.major=element_blank(),
                     panel.grid.minor=element_blank())

#separate based on DNASE overlap
#FirstlLoad DNAse data
dnase<-read.table("ENCFF001BVR.bed",stringsAsFactors=FALSE)
dnase<-GRanges(seqnames=dnase$V1,IRanges(start=dnase$V2,end=dnase$V3),score=dnase$V7,score2=dnase$V8)

dux4_vs_input_anno$dnase <- dux4_vs_input_anno %over% dnase
overs_dnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$insideFeature=="overlapStart" & dux4_vs_input_anno$dnase,]$feature)
overs_nodnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$insideFeature=="overlapStart" & !dux4_vs_input_anno$dnase,]$feature)

fivekb_dnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 5000 & dux4_vs_input_anno$dnase,]$feature)
fivekb_nodnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 5000 & !dux4_vs_input_anno$dnase,]$feature)
fivekb_dnase<-fivekb_dnase[!fivekb_dnase %in% overs]
fivekb_nodnase<-fivekb_nodnase[!fivekb_nodnase %in% overs]

tenkb_dnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 10000 & dux4_vs_input_anno$dnase,]$feature)
tenkb_nodnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 10000 & !dux4_vs_input_anno$dnase,]$feature)
tenkb_dnase<-tenkb_dnase[!(tenkb_dnase %in% fivekb) & !(tenkb_dnase %in% overs)]
tenkb_nodnase<-tenkb_nodnase[!(tenkb_nodnase %in% fivekb) & !(tenkb_nodnase %in% overs)]

fiftykb_dnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 50000 & dux4_vs_input_anno$dnase,]$feature)
fiftykb_nodnase<-unique(dux4_vs_input_anno[dux4_vs_input_anno$shortestDistance < 50000 & !dux4_vs_input_anno$dnase,]$feature)
fiftykb_dnase<-tenkb[!(fiftykb_dnase %in% tenkb) & !(fiftykb_dnase %in% fivekb) & !(fiftykb_dnase %in% overs)]
fiftykb_nodnase<-tenkb[!(fiftykb_nodnase %in% tenkb) & !(fiftykb_nodnase %in% fivekb) & !(fiftykb_nodnase %in% overs)]

res$dnase<-NA
res[rownames(res) %in% overs_dnase,"dnase"]<-TRUE
res[rownames(res) %in% fivekb_dnase,"dnase"]<-TRUE
res[rownames(res) %in% tenkb_dnase,"dnase"]<-TRUE
res[rownames(res) %in% fiftykb_dnase,"dnase"]<-TRUE
res[rownames(res) %in% overs_nodnase,"dnase"]<-FALSE
res[rownames(res) %in% fivekb_nodnase,"dnase"]<-FALSE
res[rownames(res) %in% tenkb_nodnase,"dnase"]<-FALSE
res[rownames(res) %in% fiftykb_nodnase,"dnase"]<-FALSE


gg2<-res %>% 
  dplyr::filter(peak10kb != "> 50kb" & dnase) %>% 
  mutate(mlogp=-1*log10(padj)) %>% 
  mutate(peak10kb=factor(peak10kb,levels=c("Overlap's Start","< 5Kb","< 10Kb","< 50Kb","> 50kb"))) %>% 
  ggplot(aes(x=log2FoldChange,y=mlogp,color=peak10kb)) +
  geom_point(size=3) + scale_color_manual(values=cbPalette[c(2,6,7,8,1)],
                                    name="Distance to Dux4 Peak") +
  xlab("DUX4 Induced Log2 Fold Change")+
  ggtitle("DNAse Accesible") +
  ylab("-log10(Adjusted P Value)") +
  xlim(c(-3,7)) + ylim(c(0,300)) + 
  theme_bw() + theme(panel.grid.major=element_blank(),
                     panel.grid.minor=element_blank())

gg3<-res %>% 
  dplyr::filter(peak10kb != "> 50kb" & !dnase) %>% 
  mutate(mlogp=-1*log10(padj)) %>% 
  mutate(peak10kb=factor(peak10kb,levels=c("Overlap's Start","< 5Kb","< 10Kb","< 50Kb","> 50kb"))) %>% 
  ggplot(aes(x=log2FoldChange,y=mlogp,color=peak10kb)) +
  geom_point(size=3) + scale_color_manual(values=cbPalette[c(2,6,7,8,1)],
                                    name="Distance to Dux4 Peak") +
  xlab("DUX4 Induced Log2 Fold Change")+
  ggtitle("DNAse In-accesible") +
  ylab("-log10(Adjusted P Value)") +
  xlim(c(-3,7)) + ylim(c(0,300)) + 
  theme_bw() + theme(panel.grid.major=element_blank(),
                     panel.grid.minor=element_blank())

save(gg1,gg2,gg3,file=paste0(ts,"_figure6c.rdata"))
grid.arrange(gg1,gg2,gg3,ncol=1)

#no Cairo on mesabi
#ggsave(file="010816_volcano1.svg",device = svglite::svglite,plot=gg1,width=8.5,height=6)
#ggsave(file="010816_volcano2.svg",device = svglite::svglite,plot=gg2,width=8.5,height=6)
#ggsave(file="010816_volcano3.svg",device = svglite::svglite,plot=gg3,width=8.5,height=6)

#plot(res$log2FoldChange,-1*log10(res$padj),cex=0.5,pch=16,
#     xlab="Dux4 induction Log2Fold Change",ylab="-log10(adjusted p-value)")
#identify(res$log2FoldChange,-1*log10(res$padj),labels=res$hgnc)

```

# Use the rlog to find some high variance genes that aren't likely due to noise.
```{r eval=F}
rld <- rlog(cds, blind=FALSE)
hist(res$pvalue[res$baseMean > 1], breaks=0:20/20, col="grey50", border="white")

topVarGenes <- head(order(rowVars(assay(rld)),decreasing=TRUE),50)
mat <- assay(rld)[ topVarGenes, ]
#mean center
mat <- mat - rowMeans(mat)

#or Z-scores
#mat <- (mat - rowMeans(mat))/rowSds(mat)

df <- as.data.frame(colData(rld)[,c("dox")])
colnames(df)<-"Condition"
dr <- AnnotationDbi::select(Homo.sapiens,keys=rownames(mat),keytype="ENSEMBL",columns="SYMBOL")
dr<-dr[!is.na(dr$SYMBOL),]
dr<-dr[!duplicated(dr$ENSEMBL),]

#show only genes that have symbols
colnames(mat)<-1:6
mat<-mat[rownames(mat) %in% dr$ENSEMBL,]
idx<-match(rownames(mat),dr$ENSEMBL)

pheatmap(mat, annotation_col=df,labels_row=dr[idx,"SYMBOL"])
```

#Genome Distribution
```{r eval=F}
(resGR <- results(cds, lfcThreshold=1, format="GRanges"))
resGR<-resGR[!is.na(resGR$log2FoldChange)]
resGR<-resGR[abs(resGR$log2FoldChange) > 2]
resGR<-keepSeqlevels(resGR,c(1:22,"X","Y"))
length(resGR) #1190

hg38IdeogramCyto <- getIdeogram("hg38", cytoband = TRUE)
#seqlevelsStyle(hg38IdeogramCyto)<-"NCBI"
hg38IdeogramCyto<-keepSeqlevels(hg38IdeogramCyto,c(1:22,"X","Y"))
seqlengths(hg38IdeogramCyto)

ggplot(hg38IdeogramCyto) + layout_karyogram(cytoband = FALSE)  + theme_bw() +
  layout_karyogram(resGR, geom = "rect", ylim = c(11, 21),color = "red")
```

#GOseq for enriched Terms
```{r eval=F}
res<-results(cds)
res<-as.data.frame(res)
res<-res[!is.na(res$padj),]
head(res)

gocat<-AnnotationDbi::select(Homo.sapiens,keys=rownames(res),keytype="ENSEMBL",columns="GOID")

#gocat<-gocat[!is.na(gocat$ENSEMBL),]
#sum(is.na(gocat$ENTREZID))
gocat<-gocat[gocat$ONTOLOGY=="BP",c("ENSEMBL","GOID")]
str(gocat)
gocat$GOID<-as.character(gocat$GOID)
#gocat2<-do.call( rbind, lapply( rownames(gocat)[1:20], testCategory ) )
gocat.list<-split(gocat$GOID,gocat$ENSEMBL)
gocat.list[["ENSG00000183337"]]

#bias.data
bd<-sum(width(reduce(ex)))
bd["ENSG00000183337"]
bd<-bd[names(bd) %in% rownames(res)]

degs<-as.numeric(res$padj < 0.05 & abs(res$log2FoldChange) > 1)
names(degs)<-rownames(res)
table(degs)

pwf<-nullp(degs,bias.data=bd)
GO.wall<-goseq(pwf,gene2cat=gocat.list)

head(GO.wall,20) %>%
  mutate(term=factor(term,levels=rev(term))) %>%
ggplot(aes(x=term,y=-log10(over_represented_pvalue))) +
  geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") +
  theme_bw() 

temp<-GO.wall[GO.wall$over_represented_pvalue < 0.05,1:2]
rownames(temp)<-temp$category
temp<-temp[,-1,drop=F]
View(temp)

listGO<-function(goid) {
print(OrganismDbi::select(Homo.sapiens,keys=goid,keytype="GOID",columns="TERM"))
tg<-OrganismDbi::select(Homo.sapiens,keys=gocat[grep(goid,gocat$GOID),"ENSEMBL"],keytype="ENSEMBL",columns="SYMBOL")
tg$deg<-degs[tg$ENSEMBL]
tg
}

listGO("GO:0045596")
```


#Supplemental Figure

# DNAse 
```{r dnase_Tornado,eval=F,fig.height=10,fig.width=10}
dnaseT_dnase<-tornado(dux4dox_1k_dnase,dataset=dnaseFS,pad = 3500,ord=0,window=5,color="cyan")
ggsave(file=paste0(ts,"_supfigure2_dnaseT_dnase.png"),plot=dnaseT_dnase,width=5,height=8)
dnaseT_nodnase<-tornado(dux4dox_1k_nodnase,dataset=dnaseFS,pad = 3500,ord=0,window=5,color="cyan")
ggsave(file=paste0(ts,"_supfigure2_dnaseT_nodnase.png"),plot=dnaseT_nodnase,width=5,height=8)
grid.arrange(dnaseT_dnase,dnaseT_nodnase,ncol=2)
```

# DUX4 
```{r dux4_Tornado,eval=F,fig.height=10,fig.width=10}
dux4T_dnase<-tornado(dux4dox_1k_dnase,dataset=DUX4i,pad = 3500,ord=0,window=5,color="blue")
ggsave(file=paste0(ts,"_supfigure2_dux4T_dnase.png"),plot=dux4T_dnase,width=5,height=8)
dux4T_nodnase<-tornado(dux4dox_1k_nodnase,dataset=DUX4i,pad = 3500,ord=0,window=5,color="blue")
ggsave(file=paste0(ts,"_supfigure2_dux4T_nodnase.png"),plot=dux4T_nodnase,width=5,height=8)
grid.arrange(dnaseT_dnase,dnaseT_nodnase,ncol=2)
```

# H3 
```{r h3_Tornado,eval=F,fig.height=10,fig.width=10}
H3T_dnase<-tornado(dux4dox_1k_dnase,dataset=H3,pad = 3500,ord=0,window=5,color="darkgoldenrod2")
ggsave(file=paste0(ts,"_supfigure2_H3_dnase.png"),plot=H3T_dnase,width=5,height=8)
H3T_nodnase<-tornado(dux4dox_1k_nodnase,dataset=H3,pad = 3500,ord=0,window=5,color="darkgoldenrod2")
ggsave(file=paste0(ts,"_supfigure2_H3_nodnase.png"),plot=H3T_nodnase,width=5,height=8)
grid.arrange(H3T_dnase,H3T_nodnase,ncol=2)
```

# H3K4me3 
```{r h3k4_Tornado,eval=F,fig.height=10,fig.width=10}
H3K4T_dnase<-tornado(dux4dox_1k_dnase,dataset=h3k4,pad = 3500,ord=0,window=5,color="darkorange4")
ggsave(file=paste0(ts,"_supfigure2_H3_dnase.png"),plot=H3K4T_dnase,width=5,height=8)
H3K4T_nodnase<-tornado(dux4dox_1k_nodnase,dataset=h3k4,pad = 3500,ord=0,window=5,color="darkorange4")
ggsave(file=paste0(ts,"_supfigure2_H3K4_nodnase.png"),plot=H3K4T_nodnase,width=5,height=8)
grid.arrange(H3K4T_dnase,H3K4T_nodnase,ncol=2)
```

# H3K27Ac
```{r h3k27_Tornado,eval=F,fig.height=10,fig.width=10}
H3K27T_dnase<-tornado(dux4dox_1k_dnase,dataset=h3k27,pad = 3500,ord=0,window=5,color="red2")
ggsave(file=paste0(ts,"_supfigure2_H3_dnase.png"),plot=H3K27T_dnase,width=5,height=8)
H3K27T_nodnase<-tornado(dux4dox_1k_nodnase,dataset=h3k27,pad = 3500,ord=0,window=5,color="red2")
ggsave(file=paste0(ts,"_supfigure2_H3K27_nodnase.png"),plot=H3K27T_nodnase,width=5,height=8)
grid.arrange(H3K27T_dnase,H3K27T_nodnase,ncol=2)
```

```{r sessionInfo,eval=T}
sessionInfo()
```

